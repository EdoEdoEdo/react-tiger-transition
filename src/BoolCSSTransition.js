/**
 *  Copy of the Original CSSTransition
 *
 *  https://github.com/reactjs/react-transition-group/blob/master/src/CSSTransition.js
 *
 *  The difference is it accepts 'css' bool prop
 *  true: return CSSTransition
 *  false: return Transition
 *
 *  Why?
 *
 *  In order to allow smooth back and forth between 'string' and 'object/function'
 *  types for 'transition' prop of Link component, we need a single Transition
 *  component to handle both cases. If we use something like:
 *
 *  string ? CSSTransition : Transition
 *
 *  Would not work because CSSTransition is child of Transition, inserting
 *  one more node and breaking the Link animation cycle when changing from
 *  string to object/function or vice versa.
 *
 */

import PropTypes from 'prop-types';
import React from 'react';
import { Transition } from 'react-transition-group'

import {
  addClass as addOneClass,
  removeClass as removeOneClass
} from './utils';

const addClasses = (node, classes) => node && classes && classes.split(' ').forEach(c => addOneClass(node, c));
const removeClasses = (node, classes) => node && classes && classes.split(' ').forEach(c => removeOneClass(node, c));

/**
 * @docIgnore
 *
 */
const BoolCSSTransition = (props) => {

  const appliedClasses = {
    appear: {},
    enter: {},
    exit: {},
  }

  const getClassNames = (type) => {
    const { classNames } = props;
    const isStringClassNames = typeof classNames === 'string';
    const prefix = isStringClassNames && classNames
      ? `${classNames}-`
      : '';

    let baseClassName = isStringClassNames
      ? `${prefix}${type}`
      : classNames[type]

    let activeClassName = isStringClassNames
      ? `${baseClassName}-active`
      : classNames[`${type}Active`];

    let doneClassName = isStringClassNames
      ? `${baseClassName}-done`
      : classNames[`${type}Done`];

    return {
      baseClassName,
      activeClassName,
      doneClassName
    };
  }

  const onEnter = (node, appearing) => {
    removeClassNames(node, 'exit');
    addClassNames(node, appearing ? 'appear' : 'enter', 'base');

    if (props.onEnter) {
      props.onEnter(node, appearing);
    }
  }

  const onEntering = (node, appearing) => {
    const type = appearing ? 'appear' : 'enter';
    addClassNames(node, type, 'active')

    if (props.onEntering) {
      props.onEntering(node, appearing);
    }
  }

  const onEntered = (node, appearing) => {
    const type = appearing ? 'appear' : 'enter'
    removeClassNames(node, type);
    addClassNames(node, type, 'done');

    if (props.onEntered) {
      props.onEntered(node, appearing);
    }
  }

  const onExit = (node) => {
    removeClassNames(node, 'appear');
    removeClassNames(node, 'enter');
    addClassNames(node, 'exit', 'base')

    if (props.onExit) {
      props.onExit(node);
    }
  }

  const onExiting = (node) => {
    addClassNames(node, 'exit', 'active')

    if (props.onExiting) {
      props.onExiting(node);
    }
  }

  const onExited = (node) => {
    removeClassNames(node, 'exit');
    addClassNames(node, 'exit', 'done');

    if (props.onExited) {
      props.onExited(node);
    }
  }

  function addClassNames(node, type, phase) {
    let className = getClassNames(type)[`${phase}ClassName`];

    if (type === 'appear' && phase === 'done') {
      className += ` ${getClassNames('enter').doneClassName}`;
    }

    if (phase === 'active') {
      node && node.scrollTop;
    }

    appliedClasses[type][phase] = className
    addClasses(node, className)
  }

  function removeClassNames(node, type) {
    const {
      base: baseClassName,
      active: activeClassName,
      done: doneClassName
    } = appliedClasses[type]

    appliedClasses[type] = {};

    if (baseClassName) {
      removeClasses(node, baseClassName);
    }
    if (activeClassName) {
      removeClasses(node, activeClassName);
    }
    if (doneClassName) {
      removeClasses(node, doneClassName);
    }
  }

  const { css, classNames, ...other } = props;

  return (
    <Transition
      {...other}
      onEnter={css ? onEnter : other.onEnter || {} }
      onEntered={css ? onEntered : other.onEntered || {} }
      onEntering={css ? onEntering : other.onEntering || {} }
      onExit={css ? onExit : other.onExit || {} }
      onExiting={css ? onExiting : other.onExiting || {} }
      onExited={css ? onExited : other.onExited || {} }
    />
  );
}

BoolCSSTransition.defaultProps = {
  css: false,
}

BoolCSSTransition.propTypes = {
  ...Transition.propTypes,
  onEnter: PropTypes.func,
  onEntering: PropTypes.func,
  onEntered: PropTypes.func,
  onExit: PropTypes.func,
  onExiting: PropTypes.func,
  onExited: PropTypes.func,
  css: PropTypes.bool,
};

export default BoolCSSTransition
