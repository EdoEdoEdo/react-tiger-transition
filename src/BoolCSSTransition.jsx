/**

  Copy of the Original CSSTransition

  https://github.com/reactjs/react-transition-group/blob/master/src/CSSTransition.js

  The difference is it accepts 'css' bool prop
  true: return CSSTransition
  false: return Transition

  Why?

  In order to allow smooth back and forth between 'string' and 'object/function'
  types for 'animation' prop of Link component, we need a single Transition
  component to handle both cases. If we use something like:

  string ? CSSTransition : Transition

  Would not work because CSSTransition is child of Transition, inserting
  one more node and breaking the Link animation cycle when changing from
  string to object/function or vice versa.

 */

import PropTypes from 'prop-types';
import React from 'react';
import { Transition } from 'react-transition-group'

import {
  addClass as addOneClass,
  removeClass as removeOneClass
} from './utils';

const addClass = (node, classes) => node && classes && classes.split(' ').forEach(c => addOneClass(node, c));
const removeClass = (node, classes) => node && classes && classes.split(' ').forEach(c => removeOneClass(node, c));


class BoolCSSTransition extends React.Component {

  appliedClasses = {
    appear: {},
    enter: {},
    exit: {},
  }

  onEnter = (node, appearing) => {
    this.removeClasses(node, 'exit');
    this.addClass(node, appearing ? 'appear' : 'enter', 'base');

    if (this.props.onEnter) {
      this.props.onEnter(node, appearing)
    }
  }

  onEntering = (node, appearing) => {
    const type = appearing ? 'appear' : 'enter';
    this.addClass(node, type, 'active')

    if (this.props.onEntering) {
      this.props.onEntering(node, appearing)
    }
  }

  onEntered = (node, appearing) => {
    const type = appearing ? 'appear' : 'enter'
    this.removeClasses(node, type);
    this.addClass(node, type, 'done');

    if (this.props.onEntered) {
      this.props.onEntered(node, appearing)
    }
  }

  onExit = (node) => {
    this.removeClasses(node, 'appear');
    this.removeClasses(node, 'enter');
    this.addClass(node, 'exit', 'base')

    if (this.props.onExit) {
      this.props.onExit(node)
    }
  }

  onExiting = (node) => {
    this.addClass(node, 'exit', 'active')

    if (this.props.onExiting) {
      this.props.onExiting(node)
    }
  }

  onExited = (node) => {
    this.removeClasses(node, 'exit');
    this.addClass(node, 'exit', 'done');

    if (this.props.onExited) {
      this.props.onExited(node)
    }
  }

  getClassNames = (type) => {
    const { classNames } = this.props;
    const isStringClassNames = typeof classNames === 'string';
    const prefix = isStringClassNames && classNames
      ? `${classNames}-`
      : '';

    let baseClassName = isStringClassNames
      ? `${prefix}${type}`
      : classNames[type]

    let activeClassName = isStringClassNames
      ? `${baseClassName}-active`
      : classNames[`${type}Active`];

    let doneClassName = isStringClassNames
      ? `${baseClassName}-done`
      : classNames[`${type}Done`];

    return {
      baseClassName,
      activeClassName,
      doneClassName
    };
  }

  addClass(node, type, phase) {
    let className = this.getClassNames(type)[`${phase}ClassName`];

    if (type === 'appear' && phase === 'done') {
      className += ` ${this.getClassNames('enter').doneClassName}`;
    }

    if (phase === 'active') {
      node && node.scrollTop;
    }

    this.appliedClasses[type][phase] = className
    addClass(node, className)
  }

  removeClasses(node, type) {
    const {
      base: baseClassName,
      active: activeClassName,
      done: doneClassName
    } = this.appliedClasses[type]

    this.appliedClasses[type] = {};

    if (baseClassName) {
      removeClass(node, baseClassName);
    }
    if (activeClassName) {
      removeClass(node, activeClassName);
    }
    if (doneClassName) {
      removeClass(node, doneClassName);
    }
  }

  render() {
    const { classNames: _, css, ...other } = this.props;

    return (
      <Transition
        {...other}
        onEnter={css ? this.onEnter : other.onEnter }
        onEntered={css ? this.onEntered : other.onEntered }
        onEntering={css ? this.onEntering : other.onEntering }
        onExit={css ? this.onExit : other.onExit }
        onExiting={css ? this.onExiting : other.onExiting }
        onExited={css ? this.onExited : other.onExited }
      />
    );
  }
}

BoolCSSTransition.defaultProps = {
  css: false,
}

BoolCSSTransition.propTypes = {
  ...Transition.propTypes,
  onEnter: PropTypes.func,
  onEntering: PropTypes.func,
  onEntered: PropTypes.func,
  onExit: PropTypes.func,
  onExiting: PropTypes.func,
  onExited: PropTypes.func,
  css: PropTypes.bool,
};

export default BoolCSSTransition
