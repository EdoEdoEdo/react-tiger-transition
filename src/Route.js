import React from 'react';
import PropTypes from 'prop-types';
import { Route as RouterRoute } from "react-router-dom";

import { computeClassName } from './utils';
import Transition from './Transition';

const Route = ({
  children,
  timeout,
  className,
  disableStyle,
  containerProps,
  transitionProps,
  ...other,
}) => {

  const _className = computeClassName(
    disableStyle,
    className,
    'react-tiger-transition--route'
  )

  return (
    <RouterRoute {...other}>
      {props => (
        <Transition
          match={props.match}
          timeout={timeout}
          className={_className}
          containerProps={containerProps}
          transitionProps={transitionProps}
        >
          {children}
        </Transition>
      )}
    </RouterRoute>
  )
}

Route.defaultProps = {
  timeout: 600,
  disableStyle: false,
};

Route.propTypes = {
  /**
   *  Propably your 'page' component. I recommend you to use <Screen />
   *  to wrap your pages.
   */
  children: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.node,
  ]).isRequired,

  /**
   *  Transition timeout in milliseconds. Used only on CSS transitions.
   *  If using an object or a function for 'transition' prop from
   *  <Link />, should return the
   *  timeout there.
   */
  timeout: PropTypes.number,

  /**
   *  Disable default styles applied to div container. You can
   *  still use className to set your own styles.
   */
  disableStyle: PropTypes.bool,

  /**
   *  Div container className. A string or a function returning a string.
   *  If not disableStyle, this className will be chained to
   *  'react-tiger-transition--route'.
   */
   className: PropTypes.oneOfType([
     PropTypes.string,
     PropTypes.func,
   ]),

 /**
  *  Props passed to <Transition /> or <CSSTransition /> from
  *  react-transition-group. Usually you don't need to worry about this.
  *  If you pass appear=true, the appearing animation is the default one defined
  *  in <Navigation />
  */
 transitionProps: PropTypes.object,

  /**
   *  Props passed to div container.
   */
  containerProps: PropTypes.object,
}

export default Route
