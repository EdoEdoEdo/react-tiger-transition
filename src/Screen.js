import React, { useState, useMemo, useEffect } from 'react';
import PropTypes from 'prop-types';
import { withRouter, matchPath } from 'react-router-dom';

import { computeClassName } from './utils';


const ScreenBase = React.forwardRef(({
  className,
  children,
  ...other
}, ref) => (
  <div
    ref={ref}
    className={className}
    {...other}
  >
    {children}
  </div>
));


const Display = withRouter(({
  match,
  location,
  history,
  children,
}) => {

  const [isFirstRender, setIsFirstRender] = useState(true);
  const [baseMatch, setBaseMatch] = useState(match)

  useEffect(() => {
    setIsFirstRender(false);
  }, [])

  const onDisplayPath =
    matchPath(location.pathname, {...baseMatch}) !== null ?
    true :
    false;

  const cancelAnimation = isFirstRender || !onDisplayPath;

  const computeChildren = useMemo(() => {
    let clonedChildren = [];
    let element;
    React.Children.forEach(children, child => {
      if (React.isValidElement(child)) {
        element = child;
        clonedChildren.push(React.cloneElement(element, {cancelAnimation}));
      }
    });
    return clonedChildren;
  }, [children])

  return useMemo(() => (
    <React.Fragment>
      {computeChildren}
    </React.Fragment>
  ), [computeChildren, cancelAnimation])
})


/**
 * @description
 * A div with some default css applied to it. Designed to be child of [`<route />`](/docs/route),
 * or parent if passed the `display` prop.
 *
 * @afterProps
 * \*Ref and other props are passed to div container.
 *
 * @example
 * import { Screen } from 'react-tiger-transition';
 *
 * <Screen>
 *  <MyPage />
 * </Screen>
 *
 */
const Screen = React.forwardRef(({
  className,
  disableStyle,
  container,
  display,
  children,
  ...other
}, ref) => {

  const _className = computeClassName(
    disableStyle,
    className,
    `react-tiger-transition--${container || display ? 'container' : 'screen'}`
  )

  const props = {
    className: _className,
    ref,
    ...other
  }

  return display ?
    <ScreenBase {...props}>
      <Display children={children} />
    </ScreenBase>
     :
    <ScreenBase {...props} children={children}/>
})

Screen.displayName = 'TigerScreen';

Screen.defaultProps = {
  disableStyle: false,
  container: false,
  display: false,
}

Screen.propTypes = {
  /**
   * Transform it into a container to wrap routes. `<Navigation />` already
   * does that. If you want to create a display container with transitions,
   * use the `display` props.
   */
  container: PropTypes.bool,

  /**
   * Create a display of routes. Works the same way, but does not animate
   * on arriving and leaving to the base url of the screen.
   */
  display: PropTypes.bool,

  /**
   * Disable default styles applied to div container. You can
   * still use `className` to set your own styles.
   */
  disableStyle: PropTypes.bool,

  /**
   * Div container `className`. A string or a function returning a string.
   * If not `disableStyle`, this `className` will be chained to
   * `react-tiger-transition--screen`.
   */
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
  ]),
}

export default Screen
