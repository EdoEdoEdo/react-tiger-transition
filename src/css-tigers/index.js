export { flip } from './flip';
export { glide } from './glide';
export { shuffle } from './shuffle';
