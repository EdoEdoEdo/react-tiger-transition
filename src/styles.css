body {
	overflow: hidden;
	margin: 0px;
	padding: 0px;
}

.react-tiger-transition--container {
	position: relative;
	width: 100%;
	height: 100%;
	-webkit-perspective: 1200px;
	-moz-perspective: 1200px;
	perspective: 1200px;
  overflow: hidden;
}

.react-tiger-transition--route, .react-tiger-transition--fixed {
	width: 100%;
	position: absolute;
	visibility: visible;
	overflow: hidden;
	-webkit-backface-visibility: hidden;
	-moz-backface-visibility: hidden;
	backface-visibility: hidden;
	-webkit-transform: translate3d(0, 0, 0);
	-moz-transform: translate3d(0, 0, 0);
	transform: translate3d(0, 0, 0);
	-webkit-transform-style: preserve-3d;
	-moz-transform-style: preserve-3d;
	transform-style: preserve-3d;
}

.react-tiger-transition--route {
	height: 100%;
	top: 0;
	left: 0;
  bottom: 0;
  right: 0;
}

.react-tiger-transition--fixed {
	z-index: 3;
}

.react-tiger-transition--screen {
  width: 100%;
  height: 100%;
  position: relative;
  overflow-y: scroll;
}
