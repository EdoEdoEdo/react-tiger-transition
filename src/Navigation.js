import React, { useMemo, useReducer } from 'react';
import { withRouter, matchPath, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';

import { computeClassName } from './utils';
import { fade } from './tigers';
import Screen from './Screen';

import './style.css';

export const NavigationContext = React.createContext();

const evalTransition = ({transition, timeout}) => (
  typeof(transition) === 'function' ?
  { timeout, ...transition(), css: false } :
  Object.prototype.toString.call(transition) === '[object Object]' ?
  { timeout, ...transition, css: false } :
  { timeout, classNames: transition, css: true}
)

function reducer(state, action) {
  switch (action.type) {

    case 'setTransition':
      const transition = evalTransition({...action.value});
      window.setTimeout(function() {
        action.dispatch({type: 'endTransition'})
      }, transition.timeout + 200);
      return {
        ...state,
        transition,
        onTransition: true,
      };

    case 'endTransition':
      return {
        ...state,
        onTransition: false,
      };

    default:
      return { ...state };
  }
}

const NavigationProvider = withRouter(({
  children,
  defaultRoute,
  initialState,

  match,
  location,
  history,
}) => {

  const [state, dispatch] = useReducer(reducer, initialState);

  const context = {
    ...state,
    setTransition: (transition, timeout) => new Promise(
      function(resolve, reject) {
        resolve(dispatch({
          type: 'setTransition',
          value: { transition, timeout },
          dispatch
        }))
      }
    ),
    endTransition: () => dispatch({type: 'endTransition'})
  }

  const matched = useMemo(() => (
    children.some((child, i) => {
      const matchedPath = matchPath(location.pathname, {
        path: child.props.path,
        exact: child.props.exact || false,
        strict: child.props.strict || false,
      });
      if (matchedPath != null){
        return true
      }
    })
  ), [location.pathname, children])

  return (
    <NavigationContext.Provider value={context}>
      {children}
      {!matched && defaultRoute}
    </NavigationContext.Provider>
  )
})

/**
 *
 * @description
 * The context provider for `<Link />`. Allow link to change transition on
 * click, before the routing. This makes possible to build as many different
 * transitions for the same route as possible, all evaluated on the fly.
 *
 * @example
 * import { BrowserRouter as Router} from "react-router-dom";
 *
 * import {
 *   Navigation,
 *   Route,
 *   Screen,
 *   Link,
 * } from "react-tiger-transition";
 *
 * // Don't forget to import styles!
 * import "react-tiger-transition/style.css";
 *
 * ...
 * <div id='root' style={{height: '100vh'}}>
 *   <Router>
 *     <Navigation>
 *
 *       ... my routes ...
 *
 *     </Navigation>
 *   </Router>
 * </div>
 * ...
 *
 *
 */
const Navigation = ({
  children,
  containerProps,

  defaultTransition,
  firstTimeout,
  globalTransitionProps,
  ...other,
}) => (
  <Screen container {...containerProps}>
    <NavigationProvider
      {...other}
      initialState={{
          transition: evalTransition({
            transition: defaultTransition,
            timeout: firstTimeout
          }),
          currentTransition: null,
          onTransition: false,
          defaultTransition,
          globalTransitionProps,
      }}
    >
        {children}
    </NavigationProvider>
  </Screen>
)

Navigation.defaultProps = {
  defaultTransition: fade,
  defaultRoute: <Redirect to='/' />,
  globalTransitionProps: {},
  firstTimeout: 600,
};

Navigation.propTypes = {
  /**
   * Props passed to `<Screen container />` (that wraps the routes).
   */
   containerProps: PropTypes.object,

  /**
   * The default transition to be consumed by every `<Link />` component that
   * transition prop is not specified. Good if you want the same transition for
   * all routes, or most of them. Use string if you have your own css animation,
   * or an object or function returning an object to be passed to <Transition />
   * component from react-transition-group.
   */
  defaultTransition: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.func,
  ]).isRequired,

  /**
   * Props passed to `<Transition />` or `<CSSTransition />` from
   * [react-transition-group](https://github.com/reactjs/react-transition-group).
   * Usually you don't need to worry about this.
   * If you pass `appear={true}`, the appearing animation is the defaultTransition
   * prop.
   */
  globalTransitionProps: PropTypes.object,

  /**
   * A route that matches when all routes do not. Default is
   * Redirect component from react-router-dom.
   */
  defaultRoute: PropTypes.element,

  /**
   * First transition timeout in milliseconds. Used only on appearing (if set),
   * and only if you are using a css transition. If you are using an object
   * or function returning an transition with timeout, this firstTimeout is
   * ignored.
   */
  firstTimeout: PropTypes.number,
}

export default Navigation
