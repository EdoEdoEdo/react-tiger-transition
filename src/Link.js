import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { Link as RouterLink } from "react-router-dom";

import { NavigationContext } from './Navigation';

import { withRouterAndRef } from './utils';

/**
 * @description
 * Change transitions on the fly. The next transition is evaluated before
 * the route change, allowing to go or arrive to with different transitions
 * for the same routes.
 *
 * @afterProps
 * \*Ref and other props are passed to [react router `<Link />`](https://reacttraining.com/react-router/web/api/Link).
 *
 * @example
 * import { Link, glide } from "react-tiger-transition";
 *
 * // default transition
 * <Link to='/a' transition={glide}>
 *   A
 * </Link>
 *
 * // default transition with custom args
 * <Link to='/b' transition={() => glide({...args})}>
 *   B
 * </Link>
 *
 * // your custom css classname for transitions
 * <Link to='/c' transition='my-css-animation'>
 *   C
 * </Link>
 *
 * @footer
 * \*Refer to [transitions API](/docs/transitions), for more details about
 * transitions.
 *
 */
const Link = React.forwardRef(({
  match,
  location,
  history,
  staticContext,

  transition,
  children,
  onClick,
  to,
  timeout,
  ...other,
},ref) => {

  const {
    setTransition,
    defaultTransition,
    onTransition
  } = useContext(NavigationContext)

  return (
    <RouterLink
      onClick={() => {
        if (!onTransition){
          setTransition(transition || defaultTransition, timeout)
          if (typeof onClick === 'function') onClick();
        }
      }}
      ref={ref}
      to={onTransition ? location.pathname : to}
      {...other}
    >
      {children}
    </RouterLink>
  )
})

Link.defaultProps = {
  timeout: 600,
};

Link.propTypes = {
  /**
   * Transition/animation to be applied when changing route. String if using
   * CSS animations. Object or a function returning an object with props to be
   * passed to [`<Transition />`](https://reactcommunity.org/react-transition-group/transition) or
   * [`CSSTransition />`](https://reactcommunity.org/react-transition-group/css-transition).
   * Default value comes from context, defined in [`<Navigation />`](/docs/navigation) component.
   */
  transition: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.func,
  ]),

  /**
   * Function fired on link click.
   */
  onClick: PropTypes.func,

  /**
   * Transition timeout in milliseconds. Used as fallback if not provided on
   * transition object / function. Used on css transitions.
   */
  timeout: PropTypes.number,
}

export default withRouterAndRef(Link)
