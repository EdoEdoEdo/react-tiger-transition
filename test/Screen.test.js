import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter, Router } from 'react-router-dom';
import { mount } from 'enzyme';
import { createMemoryHistory } from "history";


import  { Screen, Route, Link, Navigation } from '../src';

describe('Screen', () => {

  describe('Snapshots', () => {

    test('default props', () => {
      const component = renderer.create(<Screen  />);
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('container', () => {
      const component = renderer.create(<Screen container />);
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('display', () => {
      const component = renderer.create(
        <BrowserRouter>
          <Screen display />
        </BrowserRouter>
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('disableStyle', () => {
      const component = renderer.create(<Screen disableStyle />);
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('className', () => {
      const component = renderer.create(<Screen className='custom-classname' />);
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('pass children to SimpleScreen', () => {
      const component = renderer.create(
        <Screen>
          <div className='child' />
        </Screen>);
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('pass props to container in SimpleScreen', () => {
      const component = renderer.create(
        <Screen someProp='something'>
          <div className='child' />
        </Screen>
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('display pass cancelAnimation true to children on first render', () => {
      const component = renderer.create(
        <BrowserRouter>
          <Screen display>
            <div key='child' className='child' />
          </Screen>
        </BrowserRouter>
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('display renders only valid elements', () => {
      const component = renderer.create(
        <BrowserRouter>
          <Screen display>
            {'not a element'}
          </Screen>
        </BrowserRouter>
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

  });

  describe('display', () => {

    const mockCallback = jest.fn(cancelAnimation => cancelAnimation);

    const FakeComponent = ({
      cancelAnimation
    }) => {
      mockCallback(cancelAnimation);
      return (
        <div id='child-prop'>{cancelAnimation}</div>
      )
    };

    FakeComponent.defaultProps = {
      cancelAnimation: undefined,
    };

    const history = createMemoryHistory();

    const wrapper = mount(
      <Router history={history}>
        <Navigation>
          <Route path={['/', '/valid']}>
            <Screen display>
              <FakeComponent key='child' className='child'/>
            </Screen>
          </Route>
          <Link to='/valid'>
            valid
          </Link>
        </Navigation>
      </Router>
    );

    const validLink = wrapper.find('Link');

    it('compute cancelAnimation from true to false after first render', () => {
      expect(wrapper.find('FakeComponent').props().cancelAnimation)
      .toBe(false);
      expect(mockCallback.mock.calls.length).toBe(2);
      expect(mockCallback.mock.results[0].value).toBe(true);
      expect(mockCallback.mock.results[1].value).toBe(false);
    });

    it('keep cancelAnimation as false on base match routing', () => {
      validLink.simulate('click', { button: 0 });
      wrapper.update();
      expect(wrapper.find('FakeComponent').props().cancelAnimation)
      .toBe(false);
      expect(mockCallback.mock.calls.length).toBe(2);
    });

    it('renders only children with key', () => {
      expect(() => shallow(
        <BrowserRouter>
          <Screen display>
            <div className='child' />
          </Screen>
        </BrowserRouter>
      ).toThrowError());
    });

    it('renders only children with unique keys', () => {
      expect(() => shallow(
        <BrowserRouter>
          <Screen display>
            <div key='a' className='child' />
            <div key='a' className='child' />
          </Screen>
        </BrowserRouter>
      ).toThrowError());
    });

    // #newTest: test turn cancelAnimation to true on leaving base match path

  });


});
