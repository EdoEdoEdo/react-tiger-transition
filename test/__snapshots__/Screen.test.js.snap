// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen Snapshots className 1`] = `
<div
  className="react-tiger-transition--screen custom-classname"
/>
`;

exports[`Screen Snapshots container 1`] = `
<div
  className="react-tiger-transition--container"
/>
`;

exports[`Screen Snapshots default props 1`] = `
<div
  className="react-tiger-transition--screen"
/>
`;

exports[`Screen Snapshots disableStyle 1`] = `
<div
  className=""
/>
`;

exports[`Screen Snapshots display 1`] = `
<div
  className="react-tiger-transition--container"
/>
`;

exports[`Screen Snapshots display pass cancelAnimation true to children on first render 1`] = `
<div
  className="react-tiger-transition--container"
>
  <div
    cancelAnimation={true}
    className="child"
  />
</div>
`;

exports[`Screen Snapshots display renders only valid elements 1`] = `
<div
  className="react-tiger-transition--container"
/>
`;

exports[`Screen Snapshots pass children to SimpleScreen 1`] = `
<div
  className="react-tiger-transition--screen"
>
  <div
    className="child"
  />
</div>
`;

exports[`Screen Snapshots pass props to container in SimpleScreen 1`] = `
<div
  className="react-tiger-transition--screen"
  someProp="something"
>
  <div
    className="child"
  />
</div>
`;
